docs/python/1.6/python-appapi-devguide-5.md

```diff
@@ -67,3 +77,181 @@ def main():
 ~~~~~~
 
 Any transforms that are performed on the parallelized `Stream` object occur in parallel to the degree that is specified in the `.parallel()` function. In this example, you specified 4, which means that four channels process the data in the parallel region on the graph.
+
+
+<a id="consistent_region"></a>
+
+# Consistent region
+
+Because of business requirements, some applications require that all tuples in an application are processed at least once.
+By default any state is reset to its initial state after a processing element (PE) restart. A restart may occur due to:
+
+* a failure in the PE or its resource,
+* a explicit PE restart request,
+* a parallel region width change (IBM Streams 4.3 or later)
+
+You can use a consistent region in your stream processing applications to avoid data loss due to software or hardware failure and meet your requirements for at-least-once processing.
+
+The application or a portion of it may be configured to maintain state after a PE restart by the mechanism **consistent region**.
+A **consistent region** is a subgraph where the states of callables become consistent by processing all the tuples at least once within defined points on a stream.
+On a failure, a region is reset to its last successfully persisted state, and source callable of a region can replay any tuples submitted after the restored state.
+
+![cr_reset](/streamsx.documentation/images/python/cr_reset.png)
+
+
+The consistent region feature requires a configured [checkpoint repository](https://www.ibm.com/support/knowledgecenter/SSCRJU_5.3/com.ibm.streams.cfg.doc/doc/ibminfospherestreams-configuring-checkpoint-data-store.html) in the Streams instance and Streams applications must be submitted in [DISTRIBUTED](https://streamsxtopology.readthedocs.io/en/stable/streamsx.topology.context.html#streamsx.topology.context.ContextTypes.DISTRIBUTED) context.
+
+This section will discuss how to create stream processing applications with the consistent region feature:
+
+* [Configure a consistent region](#configure)
+* [Stateful callables](#stateful)
+* [Use SPL operators supporting consistent region](#spl_cr)
+
+
+<a id="configure"></a>
+## Configure a consistent region
+
+* [streamsx.topology.topology.Stream.set_consistent()](https://streamsxtopology.readthedocs.io/en/stable/streamsx.topology.topology.html#streamsx.topology.topology.Stream.set_consistent)
+* [ConsistentRegionConfig](https://streamsxtopology.readthedocs.io/en/stable/streamsx.topology.state.html#streamsx.topology.state.ConsistentRegionConfig)
+
+The recommended way to create a `ConsistentRegionConfig` is to call either `operator_driven()` or `periodic()`.
```

Comment:
It's worth to mention, that `operator_driven()` cannot be used when a Python callable is the begin of the CR as there is no Python API to trigger the region as we have for C++ or Java SPL operators. When SPL operators (Java or C++) are the start of a CR, the operator dictates when the region is made consistent, i.e. drained and checkpointed. Often additional operator parameters are required to control when the operator is supposed to trigger the region becoming consistent.

Reply:
note box needs to be added